File: python/produce.py
#!/usr/bin/env python
from __future__ import print_function

import json
import os
import random
import time
import sys
from kafka import KafkaProducer

KAFKA_TOPIC = os.getenv("EVENTADOR_KAFKA_TOPIC", "brewery")
KAFKA_BROKERS = os.getenv("EVENTADOR_BOOTSTRAP_SERVERS", "localhost:9092")

# Setup producer connection
producer = KafkaProducer(value_serializer=lambda v: json.dumps(v).encode('utf-8'),
                         bootstrap_servers=KAFKA_BROKERS)

print("connected to {} topic {}".format(KAFKA_BROKERS, KAFKA_TOPIC))


def get_sensor():
    """Return a random temperature between 30 and 90."""
    return random.randrange(30, 90)


def sendto_eventador(payload):
    """Add a message to the produce buffer asynchronously to be sent to Eventador."""
    try:
        producer.send(KAFKA_TOPIC, payload)
    except:
        print("unable to produce to {} topic {}".format(KAFKA_BROKERS, KAFKA_TOPIC))


payload = {}
while True:
    try:
        # run forever
        sensors = ["MashTun1", "MashTun2"]

        for sensor in sensors:
            payload = {"sensor": sensor, "temp": get_sensor()}
            sendto_eventador(payload)
            print(payload)

        # Flush the produce buffer and send to kafka
        producer.flush()
        time.sleep(3)

    except KeyboardInterrupt:
        sys.exit()

File: ruby/produce.rb
#!/usr/bin/env ruby
require "json"
require "kafka"
require "yaml"

# Kafka brokers seedlist
brokers = ENV.fetch("KAFKA_BROKERS", "localhost:9092").split(",")

# Topic
topic = ENV.fetch("KAFKA_TOPIC", "defaultsink")

### SSL configuration
use_ssl = YAML.load(ENV.fetch("KAFKA_USE_SSL", "false"))
ca_cert = ENV.fetch("KAFKA_CA_CERT", nil)
client_cert = ENV.fetch("KAFKA_CLIENT_CERT", nil)
client_key = ENV.fetch("KAFKA_CLIENT_KEY", nil)

# Kafka client id
client_id = ENV.fetch("KAFKA_CLIENT_ID", "ruby-producer")

# Start a logger
logger = Logger.new("producer.log")

if use_ssl
    # Initialize kafka using SSL
    kafka = Kafka.new(
        # Inherit the local logger
        logger: logger,
        # Broker list
        seed_brokers: brokers,
        # Client identifier
        client_id: client_id,
        # SSL bits pulled from environment variables
        ssl_ca_cert: File.read(ca_cert),
        ssl_client_cert: File.read(client_cert),
        ssl_client_cert_key: File.read(client_key),
    )
else
    # Initialize kafka using plaintext
    kafka = Kafka.new(
            logger: logger,
            seed_brokers: brokers,
            client_id: client_id,
    )
end

# Async producer for buffered messaging
async_producer = kafka.async_producer

# Sample events from our beer sensors
sensors = ["MashTun1", "MashTun2"]

puts "Producing to topic: #{topic}"

# Trap ctrl-c and request exit
Kernel.trap("SIGTERM") { exit_requested = true }

exit_requested = false
# Produce until killed
while !exit_requested
    # Push sensor temp onto buffer
    sensors.each do |sensor|
        data = JSON.dump({"sensor" => sensor, "temp" => rand(30..90)})
        async_producer.produce(data, topic: topic)
        puts "#{data}"
    end
    # Deliver the messages on the buffer
    async_producer.deliver_messages
    # sleep every 3 seconds
    sleep(3)
end
# Ensure we close the async producer
async_producer.shutdown
